### ISSUE TITLE
RecurrentLayer gradient computation fails when n_batches=1

### ISSUE THREAD

**Issue by MInner** [Lasagne/Lasagne#519]
Hi, 

Here's a simple piece of code that tries to autoregress a sequence with an RNN. I've spend some time figuring out that this it would not work with `n_batches=1` for some reason (and would throw totally uninterpretable error messages in the middle of gradient computation); if that is more of a feature rather than a bug (like "don't put everything into a single batch, you idiot!"), then it should probably throw `ValueError` when a tensor with `.shape[0] = 1` is passed :)
  
	n = 100
	dim = 10
	n_hid = 5
	seq_data = np.random.rand(n, dim)

	n_batches = 2

	X = T.dmatrix('X')
	l_in = lasagne.layers.InputLayer((None, dim), input_var=X)
	seqlen, _ = l_in.input_var.shape
	l_batched = lasagne.layers.ReshapeLayer(l_in, shape=(n_batches, seqlen/n_batches, dim))
	l_rnn = lasagne.layers.RecurrentLayer(l_batched, num_units = n_hid, grad_clipping=1)
	l_rsh = lasagne.layers.ReshapeLayer(l_rnn, shape=(-1, n_hid))
	l_dense = lasagne.layers.DenseLayer(l_rsh, nonlinearity=None, num_units=dim)
	l_out = l_dense

	pred = lasagne.layers.get_output(l_out, X)[:-1, :]
	target = X[1:, :]

	loss = T.sum((pred-target)**2)

	params = lasagne.layers.get_all_params(l_out, trainable=True)
	updates = lasagne.updates.nesterov_momentum(loss, params, learning_rate=0.001, momentum=0.09)
	step = theano.function([X], loss, updates=updates)
	for i in range(100):
	    print step(seq_data)

Here's an error I get in the process of gradient computation somewhere in `theano/tensor/type.pyc:194`:

`TypeError: ('The following error happened while compiling the node', forall_inplace,cpu,grad_of_scan_fn}(Shape_i{0}.0, Subtensor{int64:int64:int64}.0, Subtensor{int64:int64:int64}.0, Subtensor{::int64}.0, Shape_i{0}.0, hidden_to_hidden.W, InplaceDimShuffle{1,0}.0), '\n', 'Non-unit value on shape on a broadcastable dimension.', (0, 0), (True, False), 'Container name "None"')`

Thanks! :) You're (and @craffel especially) are cool, and the library itself is awesome :)

**Comment by craffel** [comment]
Hi Ben, thanks for the kind words.  I guess Theano is upset by the
reshaping, I would just use a tensor3 with n_batch = 1 if you don't want to
use batches:

n = 100
dim = 10
n_hid = 5
seq_data = np.random.rand(1, n, dim).astype(theano.config.floatX)

X = T.tensor3('X')
l_in = lasagne.layers.InputLayer((1, None, dim), input_var=X)
l_rnn = lasagne.layers.RecurrentLayer(l_in, num_units=n_hid, grad_clipping=1.)
l_rsh = lasagne.layers.ReshapeLayer(l_rnn, shape=(-1, n_hid))
l_dense = lasagne.layers.DenseLayer(l_rsh, nonlinearity=None, num_units=dim)
l_out = l_dense

pred = lasagne.layers.get_output(l_out, X)[:-1, :]
target = T.reshape(X, (n, dim))[1:, :]

loss = T.sum((pred-target)**2)

params = lasagne.layers.get_all_params(l_out, trainable=True)
updates = lasagne.updates.nesterov_momentum(loss, params,
learning_rate=0.001, momentum=0.09)
step = theano.function([X], loss, updates=updates)
for i in range(100):
    print step(seq_data)


∿

**Comment by m09** [comment]
Hi, I have the same issue with the following model:

```python
class Model(object):
    def __init__(self, voc_size, emb_size, depth=4):
        self.voc_size = voc_size
        self.emb_size = emb_size
        self.depth = depth
        self.l_in = InputLayer(shape=(None, None))
        self.l_mask = InputLayer(shape=(None, None))
        self.l_emb = EmbeddingLayer(incoming=self.l_in,
                                    input_size=self.voc_size,
                                    output_size=self.emb_size)
        incoming = self.l_emb
        for i in range(1, self.depth + 1):
            layer = LSTMLayer(incoming=incoming,
                              num_units=self.emb_size,
                              grad_clipping=5,
                              mask_input=self.l_mask)
            setattr(self, 'l_enc_words_%s' % i, layer)
            incoming = layer
        self.l_slice_words = SliceLayer(incoming, -1, 1)
        self.l_reshape = ReshapeLayer(self.l_slice_words, [1, [0], [1]])
        incoming = self.l_reshape
        for i in range(1, self.depth + 1):
            layer = LSTMLayer(incoming=incoming,
                              num_units=self.emb_size,
                              grad_clipping=5)
            setattr(self, 'l_enc_sentences_%s' % i, layer)
            incoming = layer
        self.l_slice_sentences = SliceLayer(incoming, -1, 1)
```

My problem is that I do use batches at the start of the net (up until the reshape layer), so I'm not sure if I understood the proposed fix above but here I can't set n_batch.

Any idea on how to circumvent this issue?

Thanks for the awesome library!

**Comment by f0k** [comment]
> Any idea on how to circumvent this issue?

The problem may come from the `self.l_reshape = ReshapeLayer(self.l_slice_words, [1, [0], [1]])`. If you reshape with one of the dimensions fixed to size `1`, Theano will probably infer that it's broadcastable and mess something up later (it's a Theano bug). You can try reshaping to `[-1, [0], [1]]` so Theano cannot infer that the dimension is broadcastable.

Note that instead of a `SliceLayer(incoming, -1, 1)` you can pass `only_return_final=True` to the last `LSTMLayer`'s constructor, this will be a little more efficient.

**Comment by m09** [comment]
Thanks @f0k, replacing `[1, [0], [1]]` by `[-1, [0], [1]]` fixed the problem.

I knew about `only_return_final` but didn't take the time to upgrade Lasagne to the git version so it didn't work. It's done now, thanks for the suggestion ;)

Thanks again for the great library, it's the perfect fit for my needs.

**Comment by f0k** [comment]
Requires same fix as #545. Closing to have a single open issue for this.
